rfd/0157-aws-eks-discover.md

```diff
@@ -0,0 +1,205 @@
+
+Users will have two main options for enrolling EKS clusters:
+ - Enroll specific selected clusters from the list of available ones.
+ - Set up automatic discovery and enrollment of the clusters using OIDC integration.
+
+### Enrolling Selected Clusters
+
+For the enrollment of just the selected clusters, users will be further given two choices:
+- Manually run a script in the AWS CloudShell.
+- Use EKS OIDC identity provider association to enroll a cluster with just the API.
+
+Cluster enrollment will be done through the installation of the Teleport agent using the Helm chart
+`teleport-kube-agent`. 
+
+For the AWS CloudShell, the command we generate will download a script
+that will run the Helm installation of the chart on the clusters selected by the user.
+
+For completely automated cluster enrollment through the API, the Helm Go SDK will be used -
+allowing us to avoid the need for an additional step to run Helm itself elsewhere. It would require
+running a job inside the target EKS cluster with a special image that has Helm on it, and this job
+would install the Teleport kube agent. By using the Helm Go SDK, we can run the installation directly from the Teleport
+process and don't need to maintain a new special image and EKS installation job.
+
+### Automatic Discovery and Enrollment
+
+Users will also be able to set up automatic discovery and enrollment of EKS clusters in the Discover UI.
+We will reuse the existing EKS discovery capability, although a new mode will be added to it - discovery through
+the AWS integration. EKS discovery in that mode will rely on the AWS integration, selected in the Discovery UI,
+to find clusters and then install the Helm chart on them. The algorithm will be as follows:
+- List EKS clusters through the AWS integration.
+- Cross-check to see which EKS clusters are already registered in the system.
+- Once a new EKS cluster is found, associate our OIDC provider with it.
+- Using the associated provider, generate an access token and install the kube agent through the Helm chart.
+
+Users will be able to filter EKS clusters for discovery by using labels.
+
+A dynamic discovery configuration will be created as a result of users setting up automatic EKS discovery in the Discover UI.
+This configuration can be picked up by any discovery service. In the Teleport cloud, the Discovery service will be available by default,
+meaning users will not need to perform any additional actions.
+
+### EKS Identity Provider Association
+
+AWS EKS supports a method of authentication that involves associating an OIDC identity provider with the cluster.
+Such an associated identity provider can issue access tokens (JWT) that grant access to the EKS cluster groups and users.
+Since Teleport already supports OIDC providers, we can easily use this approach. It will require slight
+modifications to the AWS OIDC integration code to allow it to issue JWTs for an EKS cluster. To associate an OIDC provider with an
```

Comment:
Let's mention what happens if the cluster already has an IdP associated with it.

Reply:
I think we have it
https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/eks#DescribeClusterOutput
https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/eks@v1.35.3/types#Cluster
https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/eks@v1.35.3/types#OIDC

Reply:
I think OIDC within describecluster is for IAM OIDC but I never confirmed it.

I also think that `eksctl` supports specifying multiple OIDC providers for a single cluster. Can you check if it works at all or it only supports a single OIDC at a time.

https://github.com/eksctl-io/eksctl/blob/16000b0d79f359588756ba00be3dec3a37bbc6aa/examples/27-oidc-provider.yaml#L14

If it only accepts a single OIDC it might be a problem :cry:

Reply:
Yep, OIDC in the describeCluster is built in AWS IAM OIDC and it's present for all EKS clusters. But right, they have separate set of API endpoints for discovering identity providers `list-identity-providers/describe-identity-provider`, so we can see what provider is associated with the cluster and if it's Teleport's.

But looks like there really can be only one provider associated, when I try to associate second provider I get an error: `An error occurred (InvalidParameterException) when calling the AssociateIdentityProviderConfig operation: Multiple Identity provider configs are not supported`